### PR TITLE
Ridvan deceptive txs

### PR DESCRIPTION
**Pull request by rozaydin**
I have verified locally, with random error behavior, and end to end. i have grown enough confidence to create the PR. 

- This PR only changes the behavior of HTTP interface.
- The wallet.js changes is, where i would like to request your most attention (the changed order of lines)

some additional related information can be accessed from: 
https://github.com/handshake-org/hsd/issues/572

PR also includes, https://github.com/handshake-org/hsd/pull/537
mainly because of the following lines:

```
const handler = () => emitted++;
node.mempool.on('tx', handler);
```

If there is no listener for the `'tx'` event, the emit statement was failing with an assertion error which i have failed to undercover the root case.

mempool.js

```
...
    async addEntry(entry, view) {
    const tx = entry.tx;

    this.trackEntry(entry, view);

    this.updateAncestors(entry, addFee);

    this.emit('tx', tx, view);
...
```

I have tested with every scenario i can think of both from the postman and from regular namebase web-app. But would be happy if you have additional way of testing in mind...

And we definitely need a traffic generator, to simulate high volume wallet traffic.

### CODE REVIEW

File: lib/wallet/nodeclient.js [hunk 1]
```diff
@@ -135,13 +135,15 @@ class NodeClient extends AsyncEmitter {
...
+   * Send a transaction. Do wait for promise.
    * @param {TX} tx
+   * @param {Boolean} sendBlocking
    * @returns {Promise}
    */
 
-  async send(tx) {
-    this.node.relay(tx);
+  async send(tx, sendBlocking) {
+      sendBlocking ? await this.node.relay(tx, sendBlocking)
```
Comment (turbomaze) [inline lib/wallet/nodeclient.js:145]: nit: only use ternaries for values, for control flow use if/else

File: lib/wallet/http.js [hunk 2]
```diff
@@ -1203,7 +1203,7 @@ class HTTP extends Server {
       res.setHeader('From-Cache', fromCache);
 
       if (broadcast) {
-        const tx = await req.wallet.sendMTX(mtx, passphrase);
+        const tx = await req.wallet.sendMTX(mtx, passphrase, true);
```
Comment (turbomaze) [inline lib/wallet/http.js:1206]: nit: for boolean flags like this I would prefer to do:

```javascript
async sendMTX(tx) {
  return this._sendMTX(tx, false);
}
async sendMTXBlocking(tx) {
  return this._sendMTX(tx, true);
}
async _sendMTX(tx, sendBlocking);
```

File: lib/wallet/http.js [hunk 3]
```diff
@@ -1236,7 +1236,7 @@ class HTTP extends Server {
         return res.json(500, {errors: errors});
 
       if (broadcast) {
-        const tx = await req.wallet.sendMTX(mtx, passphrase);
+        const tx = await req.wallet.sendMTX(mtx, passphrase, true);
```
Comment (turbomaze) [inline lib/wallet/http.js:1239]: As discussed earlier today, to start, can you use `sendMTXBlocking` specifically for low-value bids? Not for any other transaction type? Let's keep the behavior as similar as we can to how things normally work until we have more confidence in this.

**Review by turbomaze — CHANGES_REQUESTED** [review]
Seems to be merge conflicts btw

File: lib/wallet/nodeclient.js (hunk 1 above)
Comment (rozaydin) [reply lib/wallet/nodeclient.js:145]: fixed ...

File: lib/wallet/http.js (hunk 3 above)
Comment (rozaydin) [reply lib/wallet/http.js:1239]: fixed ...

File: lib/wallet/http.js (hunk 2 above)
Comment (rozaydin) [reply lib/wallet/http.js:1206]: fixed

**Comment by rozaydin** [comment]
tested end2end again, postman tested as well, contains unit testing fix to reduce number of failing tests.

File: lib/wallet/wallet.js [hunk 4]
```diff
@@ -4276,7 +4276,7 @@ class Wallet extends EventEmitter {
 
   async _send(options, passphrase) {
     const mtx = await this.createTX(options, true);
-    return this.sendMTX(mtx, passphrase);
+    return this.sendMTX(mtx, passphrase, true);
```
Comment (turbomaze) [inline lib/wallet/wallet.js:4279]: 1) We only want bids to ever be blocking so this should probably be `false`.
2) Besides that, this parameter seems unused could you do another self review and clean extra parameters like this up?

File: lib/node/fullnode.js [hunk 5]
```diff
@@ -361,16 +361,18 @@ class FullNode extends Node {
...
    */
 
-  async relay(tx) {
+  async relay(tx, sendBlocking) {
     try {
       await this.sendTX(tx);
     } catch (e) {
-      this.error(e);
+      if (sendBlocking)
+        throw e;
```
Comment (turbomaze) [inline lib/node/fullnode.js:375]: This should maintain the original behavior if sendBlocking is false

File: lib/wallet/http.js [hunk 6]
```diff
@@ -1317,7 +1317,14 @@ class HTTP extends Server {
       res.setHeader('From-Cache', fromCache);
 
       if (broadcast) {
-        const tx = await req.wallet.sendMTX(mtx, passphrase);
+        const blockingSend = (bid <= 10_000_000);
```
Comment (turbomaze) [inline lib/wallet/http.js:1320]: nit: booleans should start with `is`, `has`, `does`, etc

**Review by turbomaze — CHANGES_REQUESTED** [review]
Can you fix the deleted package-lock file too?

File: lib/node/fullnode.js (hunk 5 above)
Comment (rozaydin) [reply lib/node/fullnode.js:375]: fixed ...

File: lib/wallet/http.js (hunk 6 above)
Comment (rozaydin) [reply lib/wallet/http.js:1320]: fixed ...

File: lib/wallet/wallet.js (hunk 4 above)
Comment (rozaydin) [reply lib/wallet/wallet.js]: fixed ...

**Comment by rozaydin** [comment]
> Can you fix the deleted package-lock file too?

fixed ...

**Comment by rozaydin** [comment]
I have merged with production-namebase, run the feature test suite and resolved the package-lock.json file

**Comment by rozaydin** [comment]
re-created the null address on top of production-namebase branch contains the fix for failing batch reveal tests, i am not propagating it here ...

**Review by turbomaze — APPROVED** [review]

**Pull request merged**